File: src/main/java/com/yy/leetcode/dynamicprogramming/MinimumPathSum.java
package com.yy.leetcode.dynamicprogramming;
/**
 * @author youyouhuang
 * @create 2020-04-06
 **/

/**
 * https://leetcode-cn.com/problems/minimum-path-sum/
 * <p>
 *
 * </p>
 **/
public class MinimumPathSum {
    /**
     * dp[i][j] i*j 的网络的最小路径
     */
    public int minPathSum(int[][] grid) {
        if(grid == null ||grid.length == 0 || grid[0].length == 0){
            return 0;
        }

        int col = grid.length;
        int row = grid[0].length;
        int[][] dp = new int[col][row];
        dp[0][0] = grid[0][0];
        for (int i = 1; i < col; i++) {
            dp[i][0] = dp[i-1][0] + grid[i][0];
        }
        for (int i = 1; i < row; i++) {
            dp[0][i] = dp[0][i-1] + grid[0][i];
        }

        for (int i = 1; i < col; i++) {
            for (int j = 1; j < row; j++) {
                dp[i][j] = Math.min(dp[i][j - 1], dp[i - 1][j]) + grid[i][j];
            }
        }
        return dp[col-1][row-1];
    }
}

File: src/main/java/com/yy/leetcode/dynamicprogramming/MinimumPathSum.go
package main

//func main() {
//	grid := [][]int{
//		{1, 3, 1}, {1, 5, 1}, {4, 2, 1},
//	}
//	MinPathSum(grid)
//}

//定一个包含非负整数的 m x n 网格 grid ，请找出一条从左上角到右下角的路径，使得路径上的数字总和为最小。
//
//说明：每次只能向下或者向右移动一步。
//
//来源：力扣（LeetCode）
//链接：https://leetcode-cn.com/problems/minimum-path-sum
//著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
func MinPathSum(grid [][]int) int {
	//dp[i][j] = min(dp[i-1][j],dp[i][j-1]) + grid[i][j]

	if grid == nil {
		return 0
	}

	row, col := len(grid), len(grid[0])
	dp := make([][]int, row)
	for i := range dp {
		dp[i] = make([]int, col)
	}
	dp[0][0] = grid[0][0]

	n := 1
	for n < col {
		dp[0][n] = dp[0][n-1] + grid[0][n]
		n++
	}

	n = 1
	for n < row {
		dp[n][0] = dp[n-1][0] + grid[n][0]
		n++
	}

	for i := 1; i < row; i++ {
		for j := 1; j < col; j++ {
			dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
		}
	}

	return dp[row-1][col-1]

}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}
